### PR TITLE
WIP: New runApp/runAppLegacy APIs.

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/dart-lang/angular/issues/629.

WIP: Do **not merge**. This doesn't yet pass all test cases, and will have internal breakages as currently implemented. Early review/feedback welcome, otherwise.

Namely, this is a refactor of
* Functions that start your angular application.
* The `PlatformRef` class, which was hard to follow before, and used inconsistently.
* Some experimental functions.

### CODE REVIEW

File: angular/lib/src/bootstrap/run.dart [hunk 1]
```diff
@@ -46,6 +41,7 @@ Injector _platformInjector() {
 /// Optionally may supply a [createInjector] function in order to provide
 /// services to the root of the application:
 ///
+/// ```
 /// // Assume this file is "main.dart".
 /// import 'package:angular/angular.dart';
 /// import 'main.template.dart' as ng;
```
Comment (chalin) [inline angular/lib/src/bootstrap/run.dart]: Btw, that's a lot of boilderplate:

```dart
void main() {
  runApp(ng.HelloWorldNgFactory, createInjector: helloInjector);
}

@GenerateInjector(const [
  const ClassProvider(HelloService),
])
final InjectorFactory helloInjector = ng.helloInjector$Injector;
```

Is there any way that this could be simplified to, say:

```dart
@GenerateInjector(const [
  const ClassProvider(HelloService),
])
void main() {
  runApp(ng.HelloWorldNgFactory, createInjector: $Injector);
}
```

File: angular/lib/src/bootstrap/run.dart [hunk 2]
```diff
@@ -85,20 +81,147 @@ ComponentRef<T> runApp<T>(
...
+  });
+}
+
+const _errorMessage = 'Renamed "runAppLegacy". See doc comments for details';
+
+/// See [runAppLegacy].
+///
+/// This method only worked in AngularDart 4.x and below, and relied on the
+/// pub transformer automatically rewriting this function to [bootstrapStatic].
+/// In AngularDart 5.x+, this rewrite should be done manually.
```
Comment (chalin) [inline angular/lib/src/bootstrap/run.dart]: Maybe:

> In AngularDart 5+, this rewrite is done manually.

File: angular/lib/src/bootstrap/run.dart [hunk 3]
```diff
@@ -85,20 +81,147 @@ ComponentRef<T> runApp<T>(
...
+///
+/// @Component(
+///   selector: 'app',
+///   template: 'Hello World',
+/// )
+/// class AppComponent {}
+/// ```
+///
+/// **AFTER**:
+/// ```
```
Comment (chalin) [inline angular/lib/src/bootstrap/run.dart]: There is no `class AppComponent` in the **AFTER** example.

File: angular/lib/src/bootstrap/run.dart (hunk 1 above)
Comment (matanlurey) [reply angular/lib/src/bootstrap/run.dart]: It could, but then there is more magic.

Comment (leonsenft) [reply angular/lib/src/bootstrap/run.dart]: At the very least you'd still need the `ng` prefix on `$Injector`. But then this would rely on that being some magic top level declaration in the generated file, so what happens if you want multiple generated injectors?

I think the current approach makes it more obvious what the annotation does, and provides you the flexibility to generate multiple injectors with whatever name you desire.

File: angular/lib/src/bootstrap/run.dart [hunk 4]
```diff
@@ -85,20 +81,147 @@ ComponentRef<T> runApp<T>(
...
-  appViewUtils = unsafeCast<AppViewUtils>(appInjector.get(AppViewUtils));
-  sharedStylesHost ??= new DomSharedStylesHost(document);
   final appRef = unsafeCast<ApplicationRef>(appInjector.get(ApplicationRef));
+  appViewUtils = unsafeCast<AppViewUtils>(appInjector.get(AppViewUtils));
   return appRef.bootstrap(componentFactory, appInjector);
 }
+
+/// Starts a new AngularDart application in legacy mode, with [initReflector].
+///
+/// This method was formally known as `bootstrap` or `bootstrapStatic`, and
```
Comment (4cm4k1) [inline angular/lib/src/bootstrap/run.dart]: formally => formerly ;)

Comment (matanlurey) [reply angular/lib/src/bootstrap/run.dart]: Thanks!

**Comment by matanlurey** [comment]
Just an update, I'm helping with some Dart2 migrations, and I'll be back to complete this work, probably second-half of the week. So this is on hold for now.
